### PR TITLE
Patch index creation and deletion resilience

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

* **What is the current behavior?** (You can also link to an open issue here)
Creating and deleting indexes are non-atomic processes in Marqo consisting of two steps:
- Update the Vespa application package
- Update Marqo index metadata in Vespa
A failure may put Marqo into an intermediate state where the target index can neither be deleted nor be created.

* **What is the new behavior (if this is a feature change)?**
We don’t intend to add transaction management here, but in case of a failure, the user should be able to bring Marqo to a consistent state by repeating the call and getting a 200 response. 

Tests are added to ensure Marqo can behave as mentioned above.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
running

* **Related Python client changes** (link commit/PR here)
no
* **Related documentation changes** (link commit/PR here)
no

* **Other information**:
no

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/core/index_management/index_management.py [hunk 1]
```diff
@@ -464,8 +464,9 @@ def _remove_schema(self, app: str, name: str) -> None:
         schema_path = os.path.join(app, 'schemas', f'{name}.sd')
         if not os.path.exists(schema_path):
             logger.warn(f"Schema {name} does not exist in application package, nothing to remove")
-
-        os.remove(schema_path)
+            return None
```
Comment (farshidz) [inline src/marqo/core/index_management/index_management.py]: Return None is redundant here

File: tests/tensor_search/integ_tests/test_create_index_resilience.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+from marqo.tensor_search import tensor_search
+from marqo.tensor_search.models.add_docs_objects import AddDocsParams
+
+
+class TestCreateIndexResilience(MarqoTestCase):
+    """A test to check the resilience of the create index operation.
+
+    Index creation in Marqo is not atomic. We need to ensure that the create index operation is resilient to failures.
+    Users should at least be able to bring Marqo to a consistent state by repeating the call to create index and get a
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_create_index_resilience.py]: Create and also delete

File: tests/tensor_search/integ_tests/test_create_index_resilience.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    6. Add the new schema to the service
+    7. Deploy the new application package
+    8. Add the index settings to marqo__settings index
+    9. Add the Marqo version to the marqo__config index
+    """
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls.index_management.bootstrap_vespa()
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_create_index_resilience.py]: You don't really need this, but it's okay to have it too

File: tests/tensor_search/integ_tests/test_create_index_resilience.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+    def setUp(self) -> None:
+        self.index_name = "test_index_creation_resilience"
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_create_index_resilience.py]: Use `self.random_index_name()` instead to ensure no name clash

File: tests/tensor_search/integ_tests/test_create_index_resilience.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            # Clear up the index before test
+            tensor_search.delete_index(self.config, self.index_name)
+        except IndexNotFoundError:
+            pass
+        self.index_request = self.unstructured_marqo_index_request(name=self.index_name)
+
+    def tearDown(self) -> None:
+        try:
+            # Clear up the index after test
+            tensor_search.delete_index(self.config, self.index_name)
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_create_index_resilience.py]: Call IndexManagement class

File: tests/tensor_search/integ_tests/test_create_index_resilience.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls.index_management.bootstrap_vespa()
+
+    def setUp(self) -> None:
+        self.index_name = "test_index_creation_resilience"
+        try:
+            # Clear up the index before test
+            tensor_search.delete_index(self.config, self.index_name)
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_create_index_resilience.py]: No need for this if using a random name

File: tests/tensor_search/integ_tests/test_create_index_resilience.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+        # Ensure the index can be deleted successfully
+        tensor_search.delete_index(self.config, self.index_name)
+
+    def test_downloadApplicationPackageFail(self):
+        """Test that the create index operation is resilient to failures in downloading the application package."""
+        error = Exception("Failed to download the application package")
+        with patch("marqo.vespa.vespa_client.VespaClient.download_application", side_effect=error) \
+                as mock_create_index_by_name:
+            with self.assertRaises(Exception) as e:
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_create_index_resilience.py]: You could use a custom exception class and you won't need to verify the message, or even have a message

**Review by farshidz — CHANGES_REQUESTED** [review]
* Simplify the tests, only test failure between stages that persist a change
* Test create index when bootstrapping hasn't been done, or has been done partially

File: src/marqo/core/index_management/index_management.py (hunk 1 above)
Comment (wanliAlex) [reply src/marqo/core/index_management/index_management.py]: Done

**Review by farshidz — APPROVED** [review]

**Pull request merged**